plugins/animate/squeezimize.hpp

```diff
@@ -0,0 +1,388 @@
+                float(src_box.y - self->animation_geometry.y) / self->animation_geometry.height,
+                float((src_box.x - self->animation_geometry.x) + src_box.width) /
+                self->animation_geometry.width,
+                float((src_box.y - self->animation_geometry.y) + src_box.height) /
+                self->animation_geometry.height
+            };
+
+            const glm::vec4 target_box_pos{
+                float(self->minimize_target.x - self->animation_geometry.x) / self->animation_geometry.width,
+                float(self->minimize_target.y - self->animation_geometry.y) / self->animation_geometry.height,
+                float((self->minimize_target.x - self->animation_geometry.x) + self->minimize_target.width) /
+                self->animation_geometry.width,
+                float((self->minimize_target.y - self->animation_geometry.y) + self->minimize_target.height) /
+                self->animation_geometry.height
+            };
+
+            OpenGL::render_begin(target);
+            self->program.use(wf::TEXTURE_TYPE_RGBA);
+            self->program.uniformMatrix4f("matrix", target.get_orthographic_projection());
+            self->program.attrib_pointer("position", 2, 0, vertex_data_pos);
+            self->program.attrib_pointer("uv_in", 2, 0, vertex_data_uv);
+            self->program.uniform1i("upward", upward);
+            self->program.uniform1f("progress", progress);
+            self->program.uniform4f("src_box", src_box_pos);
+            self->program.uniform4f("target_box", target_box_pos);
+            self->program.set_active_texture(src_tex);
+            GL_CALL(glDrawArrays(GL_TRIANGLE_FAN, 0, 4));
+            OpenGL::render_end();
+        }
+    };
+
+    squeezimize_transformer(wayfire_view view,
+        wf::geometry_t minimize_target, wf::geometry_t bbox) : wf::scene::view_2d_transformer_t(view)
+    {
+        this->view = view;
+        this->minimize_target = minimize_target;
+        if (view->get_output())
+        {
+            output = view->get_output();
+            output->render->add_effect(&pre_hook, wf::OUTPUT_EFFECT_PRE);
+        }
+
+        animation_geometry.x     = std::min(bbox.x, minimize_target.x);
+        animation_geometry.y     = std::min(bbox.y, minimize_target.y);
+        animation_geometry.width =
+            std::max(std::max(std::max(bbox.width,
+                minimize_target.width),
+                (minimize_target.x + minimize_target.width) - bbox.x),
+                (bbox.x + bbox.width) - minimize_target.x);
+        animation_geometry.height =
+            std::max(std::max(std::max(bbox.height,
+                minimize_target.height),
+                (minimize_target.y + minimize_target.height) - bbox.y),
+                (bbox.y + bbox.height) - minimize_target.y);
+        OpenGL::render_begin();
+        program.compile(squeeze_vert_source, squeeze_frag_source);
```

Comment:
I wonder whether this is an expensive operation. Maybe we can store the program as custom data on core, so that we don't recompile all of the time?

Reply:
No, I actually checked this (by asking in #dri-devel). As long as shader cache is enabled, it shouldn't recompile, because it stores a hash of the shader source and returns the same program if it's in the cache.